src/Covenant/JSON.hs

```diff
@@ -0,0 +1,517 @@
+
+     Version 1 2
+   ->
+     {major: 1, minor: 2}
+
+-}
+encodeVersion :: Version -> Encoding
+encodeVersion = error "TODO: Implement encodeVersion"
+
+decodeVersion :: Value -> ASGParser Version
+decodeVersion = error "TODO: Implement decodeVersion"
+
+{- DataDeclaration & its components -}
+
+{- Encodes as a simple JSON string, e.g.
+   TyName "Foo" -> "Foo"
+-}
+encodeTyName :: TyName -> Encoding
+encodeTyName = error "TODO: Implement encodeTyName"
+
+-- NOTE: Need to validate that the string conforms w/ the Type naming rules
+decodeTyName :: Value -> ASGParser TyName
+decodeTyName = error "TODO: Implement decodeTyName"
+
+{- Encodes as a simple JSON string, e.g.
+   ConstructorName "Foo" -> "Foo"
+-}
+encodeConstructorName :: TyName -> Encoding
+encodeConstructorName = error "TODO: Implement encodeConstructorName"
+
+-- NOTE: Need to validate that the string conforms w/ the Type naming rules
+decodeConstructorName :: Value -> ASGParser ConstructorName
+decodeConstructorName = error "TODO: Implement decodeConstructorName"
+
+{- Encodes as an object. E.g.:
+
+   Constructor "Just" [IntegerT]
+   ->
+   { constructorName: "Just"
+   , constructorArgs: [...]}
+
+-}
+encodeConstructor :: Constructor AbstractTy -> Encoding
+encodeConstructor = error "TODO: Implement encodeConstructor"
+
+decodeConstructor :: Value -> ASGParser (Constructor AbstractTy)
+decodeConstructor = error "TODO: Implement decodeConstructor"
+
+{- DataEncoding encodes as a typical sum type, and will look like:
+
+   {tag: "SOP", fields: []}
+   | {tag: "PlutusData", fields: [...]}
+   | {tag: "BuiltinStrategy", fields: [...]}
+-}
+encodeDataEncoding :: DataEncoding -> Encoding
+encodeDataEncoding = error "TODO: Implement encodeDataEncoding"
+
+decodeDataEncoding :: Value -> ASGParser a
+decodeDataEncoding = error "TODO: Implement decodeDataEncoding"
+
+{- PlutusDataStrategy encodes as a typical sum type. (Omitting the 'fields' field b/c it's an enumeration)
+
+   {tag: "EnumData"}
+   | {tag: "ProductListData"}
+   | {tag: "ConstrData"}
+   | {tag: "NewtypeData"}
+
+-}
+encodePlutusDataStrategy :: PlutusDataStrategy -> Encoding
+encodePlutusDataStrategy = error "TODO: Implement encodePlutusDataStrategy"
+
+decodePlutusDataStrategy :: Value -> ASGParser PlutusDataStrategy
+decodePlutusDataStrategy = error "TODO: Implement decodePlutusDataStrategy"
+
+{- PlutusDataConstructor encodes as a typical enumeration type:
+
+  {tag: "PlutusI"}
+  | {tag: "PlutusB"}
+  | {tag: "PlutusConstr"}
+  | {tag: "PlutusList"}
+  | {tag: PlutusMap}
+
+-}
+encodePlutusDataConstructor :: PlutusDataConstructor -> Encoding
+encodePlutusDataConstructor = error "TODO: Implement encodePlutusDataConstructor"
+
+decodePlutusDataConstructor :: Value -> ASGParser PlutusDataConstructor
+decodePlutusDataConstructor = error "TODO: Implement decodePlutusDataConstructor"
+
+{- DataDeclaration AbstractTy is a bit atypical. It is a sum type, but I think it makes more sense to encode
```

Comment:
Better to describe what we do, rather than what we think. So 'It is a sum type, but we encode the argument...' or so.